sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py

```diff
@@ -841,14 +838,30 @@ def upload_blob(self, repository, data, **kwargs):
             if repository is None or data is None:
                 raise ValueError("The parameter repository and data cannot be None.")
             raise
-        return complete_upload_response_headers['Docker-Content-Digest']
+        return complete_upload_response_headers['Docker-Content-Digest'], blob_size
+
+    def _upload_blob_chunk(self, location: str, data: IO, **kwargs) -> Tuple[str, str, int]:
+        hasher = hashlib.sha256()
+        buffer = data.read(DEFAULT_CHUNK_SIZE)
+        blob_size = len(buffer)
+        while len(buffer) > 0:
+            response_headers = cast(Dict[str, str], self._client.container_registry_blob.upload_chunk(
+                location,
+                BytesIO(buffer),
+                cls=_return_response_headers,
+                **kwargs
+            ))
+            location = response_headers['Location']
+            hasher.update(buffer)
```

Comment:
The hasher dosn't keep the data in memory, right?

Reply:
Right. The hasher sequentially process each `buffer` without saving it in memory.